crates/nu-protocol/src/eval_const.rs

```diff
@@ -259,39 +250,88 @@ pub(crate) fn create_nu_constant(engine_state: &EngineState, span: Span) -> Valu
+    // <dir>/nushell/vendor/autoload for every dir in XDG_DATA_DIRS on platforms other than windows. If XDG_DATA_DIRS is not set, it falls back to <PREFIX>/share if PREFIX ends in local, or <PREFIX>/local/share:<PREFIX>/share otherwise. If PREFIX is not set, fall back to /usr/local/share:/usr/share.
+    // %PROGRAM_DATA%\nushell\vendor\autoload on windows
+
+    let into_autoload_path_fn = |mut path: PathBuf| {
+        path.push("nushell");
+        path.push("vendor");
+        path.push("autoload");
+        path
+    };
+    let mut dirs = Vec::new();
+
+    // Highest precedence for runtime env var
+    std::env::var_os("NU_VENDOR_AUTOLOAD_DIR")
+        .into_iter()
+        .map(PathBuf::from)
+        .for_each(|path| dirs.push(path));
+
+    // Next up the data dir of the current user
+    dirs::data_dir()
+        .into_iter()
+        .map(into_autoload_path_fn)
+        .for_each(|path| dirs.push(path));
+
+    // We specifically also load the XDG variant for macOS, even though dirs does not
+    #[cfg(target_os = "macos")]
+    dirs.push(into_autoload_path_fn(
+        std::env::var("XDG_DATA_HOME")
+            .ok()
+            .map(PathBuf::from)
             .unwrap_or_else(|| {
-                if cfg!(windows) {
-                    let all_user_profile = match engine_state.get_env_var("ALLUSERPROFILE") {
-                        Some(v) => format!(
-                            "{}\\nushell\\vendor\\autoload",
-                            v.coerce_string().unwrap_or("C:\\ProgramData".into())
-                        ),
-                        None => "C:\\ProgramData\\nushell\\vendor\\autoload".into(),
-                    };
-                    all_user_profile
-                } else {
-                    // In non-Windows environments, if NU_VENDOR_AUTOLOAD_DIR is not set
-                    // check to see if PREFIX env var is set, and use it as PREFIX/nushell/vendor/autoload
-                    // otherwise default to /usr/share/nushell/vendor/autoload
-                    option_env!("PREFIX").map(String::from).map_or_else(
-                        || "/usr/local/share/nushell/vendor/autoload".into(),
-                        |prefix| format!("{}/share/nushell/vendor/autoload", prefix),
-                    )
-                }
+                let mut home = dirs::home_dir();
+                home.push(".local");
+                home.push("share");
+                home
+            }),
+    ));
+
+    // Third place in terms of precedence: compile time override
+    option_env!("NU_VENDOR_AUTOLOAD_DIR")
+        .into_iter()
+        .map(PathBuf::from)
+        .for_each(|path| dirs.push(path));
+
+    // Last resort: OS data directories
+    #[cfg(target_os = "macos")]
+    dirs.push(into_autoload_path_fn(PathBuf::from(
+        "/Library/Application Support",
+    )));
+    #[cfg(not(target_os = "windows"))]
+    {
+        std::env::var_os("XDG_DATA_DIRS")
```

Comment:
I think it's possible to avoid using the `unsafe` method:
```suggestion
    #[cfg(unix)]
    {
        use std::os::unix::ffi::OsStrExt;
        
        std::env::var_os("XDG_DATA_DIRS")
```
(continued below)